### ISSUE TITLE
Expand Cases where Cudagraphs can handle mutation 

### ISSUE THREAD

**Issue by eellison** [pytorch/pytorch#103985]
### 🚀 The feature, motivation and pitch

Currently in inductor cudagraphs we [skip cudagraphs](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_fx.py#L403-L404) if there is mutation in the inductor graph. There are a couple reasons for this in the non-cudagraph tree case:

- We do an [extra warmup](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_fx.py#L403-L404
) run, so the mutations would be applied twice.
- All of the non-parameters inputs are copied over to a static location. For those inputs, the mutation would not be reflected.

To expand mutation support here, we'd only want to change the cudagraph wrapping so that on the first iteration, we warmup and run the model normally, then on the second iteration record and replay. We'd only support mutation on the [static input idxs](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_fx.py#L539). 

With cudagraph trees on the first iteration we already warmup and return the outputs, then on the second iteration record the cuda graph and replay. Thus, no double mutation would occur. Additionally, because cudagraph trees uses a single memory pool, there is less copying in of tensors to a static location. Cudagraph trees could handle mutation on both Parameter inputs, and [outputs of a prior cudagraph](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/cudagraph_trees.py#L1208). 

Concretely, here are the incremental improvements we can make:

1. In cudagraph trees, dont skip mutation if the input is a static idx in
2. Update `cudagraph_trees=False` path so that we warmup on first iteration, record and replay on second, don't skip mutation on static idxs.
3. In cudagraph trees, expand skip mutation check to see if the input is a managed tensor of the cudagraph pool. This is a bit trickier because it's may change from iteration from iteration, see [check invariants](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/cudagraph_trees.py#L1427-L1431). If a mutated input is a cudagraph managed tensor, then is not, we would need to fallback to non-cudagraph path. 

We are discussing other changes that would exercise this path more often, like compiling the optimizer or putting the mutating epilogue in the inductor graph in training, so this is important to add support for within cudagraphs.
 

cc @mcarilli @ezyang

**Comment by eellison** [comment]
Fixed with https://github.com/pytorch/pytorch/pull/106098
